httpd/handler.go

```diff
@@ -340,11 +317,153 @@ func httpResults(w http.ResponseWriter, results Results) {
 
 // httpError writes an error to the client in a standard format.
 func httpError(w http.ResponseWriter, error string, code int) {
-	results := Results{
-		&Result{Err: errors.New(error)},
-	}
+	var results influxdb.Results
+	results = append(results, &influxdb.Result{Err: errors.New(error)})
 
 	w.Header().Add("content-type", "application/json")
 	w.WriteHeader(code)
 	_ = json.NewEncoder(w).Encode(results)
 }
+
+// Filters and filter helpers
+
+// parseCredentials returns the username and password encoded in
+// a request. The credentials may be present as URL query params, or as
+// a Basic Authentication header.
+// as params: http://127.0.0.1/query?u=username&p=password
+// as basic auth: http://username:password@127.0.0.1
+func parseCredentials(r *http.Request) (string, string, error) {
+	q := r.URL.Query()
+	username, password := q.Get("u"), q.Get("p")
+	if username != "" && password != "" {
+		return username, password, nil
+	}
+	auth := r.Header.Get("Authorization")
+	if auth == "" {
+		return "", "", nil
+	}
+	fields := strings.Split(auth, " ")
+	if len(fields) != 2 {
+		return "", "", fmt.Errorf("invalid Basic Authentication header")
+	}
+	bs, err := base64.StdEncoding.DecodeString(fields[1])
+	if err != nil {
+		return "", "", fmt.Errorf("invalid Base64 encoding")
+	}
+	fields = strings.Split(string(bs), ":")
+	if len(fields) != 2 {
+		return "", "", fmt.Errorf("invalid Basic Authentication value")
+	}
+	return fields[0], fields[1], nil
+}
+
+// authorize wraps a handler and ensures that if user credentials are passed in
+// an attempt is made to authenticate that user. If authentication fails, an error is returned.
+//
+// There is one exception: if there are no users in the system, authentication is not required. This
+// is to facilitate bootstrapping of a system with authentication enabled.
+func authorize(inner func(http.ResponseWriter, *http.Request, *influxdb.User), h *Handler, requireAuthentication bool) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var user *influxdb.User
+
+		// TODO corylanou: never allow this in the future without users
+		if requireAuthentication && h.server.UserCount() > 0 {
+			username, password, err := parseCredentials(r)
+			if err != nil {
+				httpError(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+			if username == "" {
+				httpError(w, "username required", http.StatusUnauthorized)
+				return
+			}
+
+			user, err = h.server.Authenticate(username, password)
+			if err != nil {
+				httpError(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+		}
+		inner(w, r, user)
+	})
+}
+
+func versionHeader(inner http.Handler, version string) http.Handler {
```

Comment:
@otoolep can you expand on this?  I'm not sure I know what you are asking (Mostly because I haven't used Godoc for past projects and I'm not up to speed on it's conventions yet).